src/devices/machine/i82586.h

```diff
@@ -9,28 +9,254 @@
+		RB_EL   = 0x8000  // end of rbd list
+	};
+
 	template <class Object> static devcb_base &static_set_out_irq_callback(device_t &device, Object &&cb) { return downcast<i82586_base_device &>(device).m_out_irq.set_callback(std::forward<Object>(cb)); }
 
 	DECLARE_WRITE_LINE_MEMBER(ca);
+	virtual void recv_cb(u8 *buf, int length) override;
 
 protected:
-	i82586_base_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, u8 datawidth, u8 addrwidth);
+	i82586_base_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, endianness_t endian, u8 datawidth, u8 addrwidth);
 
+	// standard device_* overrides
 	virtual void device_start() override;
 	virtual void device_reset() override;
 
+	virtual void device_timer(emu_timer &timer, device_timer_id id, int param, void *ptr) override;
 	virtual space_config_vector memory_space_config() const override;
 
+	// setup and scb
+	virtual void initialise() = 0;
+	virtual void process_scb();
+	virtual void update_scb();
+
+	// command unit
+	virtual void cu_execute();
+	virtual bool cu_iasetup() = 0;
+	virtual bool cu_configure() = 0;
+	virtual bool cu_mcsetup() = 0;
+	virtual bool cu_transmit(u32 command) = 0;
+	virtual bool cu_tdreflect() = 0;
+	virtual bool cu_dump() = 0;
+
+	// receive unit
+	virtual bool accept_filter(u8 *mac);
+	virtual void ru_execute(u8 *buf, int length) = 0;
+
+	// helpers
+	void set_irq(bool irq);
+	virtual u32 address(u32 base, int offset, int address) = 0;
+	virtual u32 address(u32 base, int offset, int address, u16 empty) = 0;
+	static u32 compute_crc(u8 *buf, int length, bool crc32);
+	int fetch_bytes(u8 *buf, u32 src, int length);
+	int store_bytes(u32 dst, u8 *buf, int length);
+	void dump_bytes(u8 *buf, int length);
+
+	// device_* members
 	const address_space_config m_space_config;
 	address_space *m_space;
 
 	devcb_write_line m_out_irq;
+	static const device_timer_id CU_TIMER = 0;
+	static const device_timer_id RU_TIMER = 1;
+	emu_timer *m_cu_timer;
+	emu_timer *m_ru_timer;
+
+	// interrupt state
+	bool m_cx;          // command executed (with interrupt)
+	bool m_fr;          // frame received
+	bool m_cna;         // command unit became inactive
+	bool m_rnr;         // receive unit became not ready
+	bool m_irq_state;
+	bool m_initialised;
+
+	// receive/command unit state
+	cu_state m_cu_state; // command unit state
+	ru_state m_ru_state; // receive unit state
+	u32 m_scp_address;   // system configuration pointer
+	u32 m_scb_base;      // system control block base (segmented modes)
+	u32 m_scb_address;   // system control block address (linear mode)
+	u32 m_scb_cs;        // current scb command and status
+	u32 m_cba;           // current command block address
+	u32 m_rfd;           // current receive frame descriptor address
+
+	// addresses
+	u64 m_mac;                    // current individual address
+	std::vector<u64> m_mac_multi; // multicast address filter list
```

Comment:
I don't have any answer for that. Doing it using a dumb array would be problematic from the storage size required. The device itself possibly optimises this by computing hashes and storing those instead of the raw addresses. Open to suggestions?